src/editable_tree/mod.rs

```diff
@@ -15,13 +15,137 @@ pub enum Direction {
+pub enum EditErr {
+    /* MISC ERRORS */
+    /// The keystrokes that were attempted to be executed where invalid
+    Invalid(String),
+
+    /* MOVEMENT ERRORS */
+    /// Trying to move to the child of a node with no children
+    MoveToNonexistentChild,
+    /// Trying to move to the parent of the root
+    MoveToParentOfRoot,
+    /// Trying to move to a sibling that doesn't exist
+    MoveToNonexistentSibling,
+    /// Trying to move to a sibling of the root
+    MoveToSiblingOfRoot,
+
+    /* EDITING ERRORS */
+    /// Trying to undo the earliest change
+    NoChangesToUndo,
+    /// Trying to redo the latest change
+    NoChangesToRedo,
+    /// Trying to replace a node with one that cannot be a child of its parent
+    CannotReplace(char),
+    /// Trying to insert a node that cannot be a child of the cursor
+    CannotInsertInto { c: char, parent_name: String },
+    /// Trying to add a sibling to the root
+    AddSiblingToRoot,
+    /// Trying to delete the root
+    DeletingRoot,
+}
+
+impl EditErr {
+    /// Writes an warning message of the encountered error using either `warn!` or `error!`,
+    /// depending on the severity of the error
+    fn log_message(self) {
+        match self {
+            EditErr::Invalid(keys) => log::warn!("Invalid command '{}'", keys),
+            EditErr::MoveToNonexistentChild => {
+                log::warn!("Can't move down if the cursor has no children.")
+            }
+            EditErr::MoveToNonexistentSibling => {
+                log::warn!("Can't move to a non-existent sibling.")
+            }
+            EditErr::MoveToParentOfRoot => log::warn!("Can't move to the parent of the root."),
+            EditErr::MoveToSiblingOfRoot => log::warn!("Can't move to a sibling of the root."),
+            EditErr::NoChangesToUndo => log::warn!("No changes to undo."),
+            EditErr::NoChangesToRedo => log::warn!("No changes to redo."),
+            EditErr::CannotReplace(c) => log::warn!("Can't replace cursor with '{}'", c),
+            EditErr::CannotInsertInto { c, parent_name } => {
+                log::warn!("Can't insert '{}' into {}", c, parent_name)
+            }
+            EditErr::AddSiblingToRoot => log::warn!("Can't add siblings to the root."),
+            EditErr::DeletingRoot => log::warn!("Can't delete the root."),
+        }
+    }
+}
+
+/// A [`Result`]-like type that is returned by any of the 'edit' methods in [`DAG`].
+pub enum EditResult {
+    /// The method successfully made the edit
+    Ok(EditSuccess),
+    /// The edit was somehow invalid, and the [`EditErr`] explains how
+    Err(EditErr),
+}
+
+impl EditResult {
+    /// Consumes this `EditResult` and logs an appropriate summary report (using `info!` for
+    /// [`EditOk`]s and `warn!` or `error!` for [`EditErr`]s).
+    pub fn log_message(self) {
+        match self {
+            EditResult::Ok(ok) => ok.log_message(),
+            EditResult::Err(err) => err.log_message(),
+        }
+    }
+}
```

Comment:
If `EditResult` becomes type alias, this needs to be an extension trait. This is normal, probably even idiomatic.

Reply:
Yup this works fine...